cmd/fs-v1.go

```diff
@@ -375,91 +366,63 @@ func (fs fsObjects) CopyObject(srcBucket, srcObject, dstBucket, dstObject string
 // length indicates the total length of the object.
-func (fs fsObjects) GetObject(bucket, object string, offset int64, length int64, writer io.Writer) (err error) {
+func (fs fsObjects) GetObject(bucket, object string, startOffset int64, endOffset int64) (objReader io.ReadCloser, objInfo ObjectInfo, err error) {
 	if err = checkGetObjArgs(bucket, object); err != nil {
-		return err
+		return nil, ObjectInfo{}, err
 	}
 
 	if _, err = fs.statBucketDir(bucket); err != nil {
-		return toObjectErr(err, bucket)
-	}
-
-	// Offset cannot be negative.
-	if offset < 0 {
-		return toObjectErr(traceError(errUnexpected), bucket, object)
+		return nil, ObjectInfo{}, toObjectErr(err, bucket)
 	}
 
-	// Writer cannot be nil.
-	if writer == nil {
-		return toObjectErr(traceError(errUnexpected), bucket, object)
-	}
+	fsMeta := fsMetaV1{}
 
 	if bucket != minioMetaBucket {
 		fsMetaPath := pathJoin(fs.fsPath, minioMetaBucket, bucketMetaPrefix, bucket, object, fsMetaJSONFile)
-		_, err = fs.rwPool.Open(fsMetaPath)
+		rlk, err := fs.rwPool.Open(fsMetaPath)
+		if err == nil {
+			defer fs.rwPool.Close(fsMetaPath)
+			if _, rerr := fsMeta.ReadFrom(rlk.LockedFile); rerr != nil {
+				// `fs.json` can be empty due to previously failed
+				// PutObject() transaction, if we arrive at such
+				// a situation we just ignore and continue.
+				if errorCause(rerr) != io.EOF {
+					return nil, ObjectInfo{}, toObjectErr(rerr, bucket, object)
+				}
+			}
+		}
 		if err != nil && err != errFileNotFound {
-			return toObjectErr(traceError(err), bucket, object)
+			return nil, ObjectInfo{}, toObjectErr(traceError(err), bucket, object)
 		}
-		defer fs.rwPool.Close(fsMetaPath)
 	}
 
-	// Read the object, doesn't exist returns an s3 compatible error.
-	fsObjPath := pathJoin(fs.fsPath, bucket, object)
-	reader, size, err := fsOpenFile(fsObjPath, offset)
+	objInfo, err = fs.getObjectInfo(bucket, object, fsMeta)
 	if err != nil {
-		return toObjectErr(err, bucket, object)
+		return nil, ObjectInfo{}, toObjectErr(err, bucket, object)
 	}
-	defer reader.Close()
 
-	bufSize := int64(readSizeV1)
-	if length > 0 && bufSize > length {
-		bufSize = length
+	offset, length, err := calcRange(startOffset, endOffset, objInfo.Size)
+	if err != nil {
+		return nil, ObjectInfo{}, toObjectErr(err, bucket, object)
 	}
 
-	// For negative length we read everything.
-	if length < 0 {
-		length = size - offset
+	// Read the object, doesn't exist returns an s3 compatible error.
+	fsObjPath := pathJoin(fs.fsPath, bucket, object)
+	reader, _, err := fsOpenFile(fsObjPath, offset)
+	if err != nil {
+		return nil, ObjectInfo{}, toObjectErr(err, bucket, object)
 	}
 
-	// Reply back invalid range if the input offset and length fall out of range.
-	if offset > size || offset+length > size {
-		return traceError(InvalidRange{offset, length, size})
+	objReader = reader
+	if length >= 0 {
+		objReader = ioutil.NopCloser(io.LimitReader(reader, length))
```

Comment:
Don't use `NopCloser` it wouldn't close the underlying Close() on the 'reader' leading to fd leak.